tools/system_libs.py

```diff
@@ -1978,7 +1967,9 @@ def get_ports(settings):
   ret = []
 
   process_dependencies(settings)
-  for port in ports.ports:
+  needed = set(p for p in ports.ports if p.needed(settings))
```

Comment:
i think this line is not needed, and `needed = ` can be on the previous line?

perhaps `process_dependencies` would be `get_needed_ports` as that is now what it does, process and get the needed ones, IIUC?